### PR TITLE
ZK-3507: Shift + up key can deselect listitems

### PR DESCRIPTION
**Pull request by bob31231**
this feature will be applied on both listbox and tree

### CODE REVIEW

File: zul/src/archive/web/js/zul/sel/ItemWidget.js [hunk 1]
```diff
@@ -303,6 +303,13 @@ zul.sel.ItemWidget = zk.$extends(zul.Widget, {
 	},
 	deferRedrawHTML_: function (out) {
 		out.push('<tr', this.domAttrs_({domClass: 1}), ' class="z-renderdefer"></tr>');
+	},
+	compareRowPos: function (item) {
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/sel/ItemWidget.js]: Make it protected is more suitable like `isStripeable_()` and write some jsdoc.

File: zul/src/archive/web/js/zul/sel/Treeitem.js [hunk 2]
```diff
@@ -86,6 +86,47 @@ it will be useful, but WITHOUT ANY WARRANTY.
...
+		while (true) {
+			if (thisPath[depth] < itemPath[depth]) {
+				return -1;
+			} else if (thisPath[depth] > itemPath[depth]) {
+				return 1;
+			} else if (thisPath[depth] == itemPath[depth]) {
+				if (thisPath[depth] == undefined) //just in case, it should never be run into this line.
+					break;
+				{
+				}
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/sel/Treeitem.js]: Strange { }

File: zul/src/archive/web/js/zul/sel/SelectWidget.js [hunk 3]
```diff
@@ -786,6 +797,10 @@ zul.sel.SelectWidget = zk.$extends(zul.mesh.MeshWidget, {
 
 		return _afterChildKey(evt);
 	},
+	_getToSelFlag: function (row, startRow, step) { // F85-ZK-3507
+		var pos = startRow.compareRowPos(row);
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/sel/SelectWidget.js]: What if this._startRow is null? Could throw an exception.

File: zul/src/archive/web/js/zul/sel/Treeitem.js (hunk 2 above)
Comment (bob31231) [inline zul/src/archive/web/js/zul/sel/Treeitem.js]: fixed

**Pull request merged**